File: sorting/sorting.php
#!/usr/bin/php5
<?php

//--- heapsort: in-place implementation with linear-time heap construction

function trickleDown(&$data, $pos, $n) {
	$left = 2*$pos+1;
	if ($left >= $n) {
		return;
	}
	$right = 2*$pos+2;
	if ($right >= $n) {
		if ($data[$pos] < $data[$left]) {
			$t = $data[$pos];
			$data[$pos] = $data[$left];
			$data[$left] = $t;
		}
		return;
	}
	if ($data[$pos] < $data[$left]) {
		if ($data[$left] < $data[$right]) {
			$t = $data[$pos];
			$data[$pos] = $data[$right];
			$data[$right] = $t;
			trickleDown($data, $right, $n);
		} else {
			$t = $data[$pos];
			$data[$pos] = $data[$left];
			$data[$left] = $t;
			trickleDown($data, $left, $n);
		}
	} else if ($data[$pos] < $data[$right]) {
		$t = $data[$pos];
		$data[$pos] = $data[$right];
		$data[$right] = $t;
		trickleDown($data, $right, $n);
	}
}

function heapsort(&$data) {
	# build a max-heap
	$n = count($data);
	for ($i = (int)(($n-2)/2); $i != -1; --$i) {
		trickleDown($data, $i, $n);
	}

	# successively remove the max item and swap it to the end
	for ($i = $n-1; $i != -1; --$i) {
		$t = $data[$i];
		$data[$i] = $data[0];
		$data[0] = $t;
		trickleDown($data, 0, $i);
	}
}

function testHeapsort($data) {
	timeFunction("ran heapsort", function () use (&$data) {
		heapsort($data);
	});
	if (!isSorted($data)) {
		printf("error: data is not sorted\n");
	}
}


//--- mergesort: recursive implementation with two arrays

function mergesort(&$data, &$buffer, $left, $right) {
	if ($left+1 >= $right) {
		return;
	}
	$mid = (int)(($left+$right)/2);
	mergesort($data, $buffer, $left, $mid);
	mergesort($data, $buffer, $mid, $right);
	$pos_left = $left;
	$pos_right = $mid;
	$pos_buffer = $left;
	while (true) {
		if ($data[$pos_left] <= $data[$pos_right]) {
			$buffer[$pos_buffer] = $data[$pos_left];
			$pos_buffer += 1;
			$pos_left += 1;
			if ($pos_left == $mid) {
				while ($pos_right != $right) {
					$buffer[$pos_buffer] = $data[$pos_right];
					++$pos_buffer;
					++$pos_right;
				}
				break;
			}
		} else {
			$buffer[$pos_buffer] = $data[$pos_right];
			$pos_buffer += 1;
			$pos_right += 1;
			if ($pos_right == $right) {
				while ($pos_left != $mid) {
					$buffer[$pos_buffer] = $data[$pos_left];
					++$pos_buffer;
					++$pos_left;
				}
				break;
			}
		}
	}
	for ($i = $left; $i < $right; ++$i) {
		$data[$i] = $buffer[$i];
	}
}

function testMergesort($data) {
	timeFunction("ran mergesort", function () use (&$data) {
		$buffer = [];
		mergesort($data, $buffer, 0, count($data));
	});
	if (!isSorted($data)) {
		printf("error: data is not sorted\n");
	}
}


//--- quicksort: in-place implementation with median-of-three pivot selection

function quicksort(&$data, $left, $right) {
	if ($left+1 >= $right) {
		return;
	}
	$ai = $left;
	$bi = (int)(($left+$right)/2);
	$ci = $right-1;
	$a = $data[$ai];
	$b = $data[$bi];
	$c = $data[$ci];
    if ($a < $b) {
        if ($c < $a) {
            $pos = $ai;
		} else if ($c < $b) {
			$pos = $ci;
		} else {
			$pos = $bi;
		}
	} else {
        if ($c < $b) {
            $pos = $bi;
		} else if ($c < $a) {
			$pos = $ci;
		} else {
			$pos = $ai;
		}
	}
    $pivot = $data[$pos];
	$data[$pos] = $data[$right-1];
	$tail = $left;
	for ($i = $left; $i != $right-1; ++$i) {
		if ($data[$i] < $pivot) {
			$t = $data[$tail];
			$data[$tail] = $data[$i];
			$data[$i] = $t;
			++$tail;
		}
	}
	$data[$right-1] = $data[$tail];
	$data[$tail] = $pivot;
    quicksort($data, $left, $tail);
    quicksort($data, $tail+1, $right);
}

function testQuicksort($data) {
	timeFunction("ran quicksort", function () use (&$data) {
		quicksort($data, 0, count($data));
	});
	if (!isSorted($data)) {
		printf("error: data is not sorted\n");
	}
}

function quicksortCustom(&$data, $left, $right, $compare) {
	if ($left+1 >= $right) {
		return;
	}
	$ai = $left;
	$bi = (int)(($left+$right)/2);
	$ci = $right-1;
	$a = $data[$ai];
	$b = $data[$bi];
	$c = $data[$ci];
    if ($compare($a, $b) == -1) {
        if ($compare($c, $a) == -1) {
            $pos = $ai;
		} else if ($compare($c, $b) == -1) {
			$pos = $ci;
		} else {
			$pos = $bi;
		}
	} else {
        if ($compare($c, $b) == -1) {
            $pos = $bi;
		} else if ($compare($c, $a) == -1) {
			$pos = $ci;
		} else {
			$pos = $ai;
		}
	}
    $pivot = $data[$pos];
	$data[$pos] = $data[$right-1];
	$tail = $left;
	for ($i = $left; $i != $right-1; ++$i) {
		if ($compare($data[$i], $pivot) == -1) {
			$t = $data[$tail];
			$data[$tail] = $data[$i];
			$data[$i] = $t;
			++$tail;
		}
	}
	$data[$right-1] = $data[$tail];
	$data[$tail] = $pivot;
    quicksortCustom($data, $left, $tail, $compare);
    quicksortCustom($data, $tail+1, $right, $compare);
}

function testQuicksortCustom($data, $compare) {
	timeFunction("ran quicksort", function () use (&$data, $compare) {
		quicksortCustom($data, 0, count($data), $compare);
	});
	if (!isSortedCustom($data, $compare)) {
		printf("error: data is not sorted\n");
	}
}


//--- native sort

function testNativeSortInteger($data) {
	timeFunction("ran native sort", function () use (&$data) {
		sort($data, SORT_NUMERIC);
	});
	if (!isSorted($data)) {
		printf("error: data is not sorted\n");
	}
}

function testNativeSortString($data) {
	timeFunction("ran native sort", function () use (&$data) {
		sort($data, SORT_STRING);
	});
	if (!isSorted($data)) {
		printf("error: data is not sorted\n");
	}
}

function testNativeSortCustom($data, $compare) {
	timeFunction("ran native sort", function () use (&$data, $compare) {
		usort($data, $compare);
	});
	if (!isSortedCustom($data, $compare)) {
		printf("error: data is not sorted\n");
	}
}


//--- random permutation

function shuffleData(&$data) {
    // constants for a linear congruential generator
    $m = (1 << 32);
	$a = 1664525;
	$c = 1013904223;
	$randomValue = 42;  // seed
    for ($i = count($data)-1; $i != 0; --$i) {
        $randomValue = ($a*$randomValue + $c) % $m;
        $j = $randomValue % ($i+1);
        if ($j != $i) {
            $t = $data[$i];
            $data[$i] = $data[$j];
            $data[$j] = $t;
        }
    }
}


//--- helper functions

function timeFunction($description, $function) {
	$start_time = microtime(true);
	$function();
	$elapsed = microtime(true) - $start_time;
	printf("    %s  %.3f s\n", $description, $elapsed);
}

function isSorted(&$arr) {
	$n = count($arr);
	for ($i = 1; $i < $n; ++$i) {
		if ($arr[$i-1] > $arr) {
			return false;
		}
	}
	return true;
}

function isSortedCustom(&$arr, $compare) {
	$n = count($arr);
	for ($i = 1; $i < $n; ++$i) {
		if ($compare($arr[$i-1], $arr[$i]) == 1) {
			return false;
		}
	}
	return true;
}

function display(&$arr) {
	echo $arr[0];
	$n = count($arr);
	for ($i = 1; $i < $n; ++$i) {
		echo " ".$arr[$i];
	}
	echo "\n";
}

function displayStudents(&$arr) {
	echo implode(" ", [$arr[0]->grade, $arr[0]->name]);
	$n = count($arr);
	for ($i = 1; $i < $n; ++$i) {
		echo " ".implode(" ", [$arr[$i]->grade, $arr[$i]->name]);
	}
	echo "\n";
}


function main() {
	global $argc, $argv;
	if ($argc != 2) {
		exit("error: must specify the length of the array\n");
	}
	$n = intval($argv[1]);


	//--- integer sorting
	echo "integers:\n";

    // initialize an array with 0, ..., n-1
	$sorted_integers = null;
	timeFunction("made ordered integer array", function ()
			use (&$sorted_integers, $n) {
		$sorted_integers = [];
		for ($i = 0; $i < $n; ++$i) {
			$sorted_integers[] = $i;
		}
	});

	// permute randomly
	$shuffled_integers = $sorted_integers;
	timeFunction("shuffled", function () use (&$shuffled_integers) {
		shuffleData($shuffled_integers);
	});

	// sort integer arrays
	testHeapsort($shuffled_integers);
	testMergesort($shuffled_integers);
	testQuicksort($shuffled_integers);
	testNativeSortInteger($shuffled_integers);


	//--- string sorting
	echo "strings:\n";
	
    // make strings: 'a', 'b', ..., 'z', 'aa', 'ab', ..., 'az', 'ba', ...
	$sorted_strings = null;
	timeFunction("made ordered string array", function ()
			use (&$sorted_strings, $n) {
		$sorted_strings = [];
		$next_char = [];
		$zCode = ord('z');
		for ($code = ord('a'); $code != $zCode; ++$code) {
			$next_char[chr($code)] = chr($code+1);
		}
		$chars = ['a'];
		$num_chars = 1;
		for ($i = 0; $i != $n; ++$i) {
			$sorted_strings[] = implode($chars);
			$pos = $num_chars-1;
			while (true) {
				if ($chars[$pos] != 'z') {
					$chars[$pos] = $next_char[$chars[$pos]];
					break;
				}
				$pos -= 1;
				if ($pos == -1) {
					array_unshift($chars, 'a');
					$num_chars += 1;
					break;
				}
			}
			if ($pos != $num_chars-1) {
				for ($j = $pos+1; $j <= $num_chars-1; ++$j) {
					$chars[$j] = 'a';
				}
			}
		}
	});

    // permute randomly
	$shuffled_strings = $sorted_strings;
	timeFunction("shuffled", function () use (&$shuffled_strings) {
		shuffleData($shuffled_strings);
	});

    // sort string arrays
    testQuicksort($shuffled_strings);
    testNativeSortString($shuffled_strings);


	//--- object sorting
	echo "objects:\n";

	class Student {
		function Student($name, $grade) {
			$this->name = $name;
			$this->grade = $grade;
		}
	}

	// make student objects based on the sorted strings
	$sorted_students = null;
	timeFunction("made student objects", function ()
			use (&$sorted_students, &$sorted_strings, $n) {
		$sorted_students = [];
		for ($i = 0; $i < $n; ++$i) {
			$sorted_students[] = new Student($sorted_strings[$i], $i%10);
		}
	});

    // permute randomly
    $shuffled_students = $sorted_students;
	timeFunction("shuffled", function () use (&$shuffled_students) {
		shuffle($shuffled_students);
	});

	// sort objects
	$compareStudents = function ($a, $b) {
		if ($a->grade < $b->grade) {
			return -1;
		} else if ($a->grade > $b->grade) {
			return 1;
		} else if ($a->name < $b->name) {
			return -1;
		} else if ($a->name > $b->name) {
			return 1;
		}
		return 0;
	};
    testQuicksortCustom($shuffled_students, $compareStudents);
    testNativeSortCustom($shuffled_students, $compareStudents);
}

main();

File: sorting/sorting.go
package main

import (
	"os"
	"fmt"
	"strconv"
	"time"
	"sort"
)


//--- heapsort: in-place implementation with linear-time heap construction

func trickleDown(data []int, pos, n int) {
	left := 2*pos+1
	if (left >= n) {
		return
	}
	right := 2*pos+2
	if (right >= n) {
		if (data[pos] < data[left]) {
			data[pos], data[left] = data[left], data[pos]
		}
		return;
	}
	if (data[pos] < data[left]) {
		if (data[left] < data[right]) {
			data[pos], data[right] = data[right], data[pos]
			trickleDown(data, right, n)
		} else {
			data[pos], data[left] = data[left], data[pos]
			trickleDown(data, left, n)
		}
	} else if(data[pos] < data[right]) {
		data[pos], data[right] = data[right], data[pos]
		trickleDown(data, right, n)
	}
}

func heapsort(data []int) {
	// build a max-heap
	n := len(data)
	for i := (n-2)/2; i != -1; i-- {
		trickleDown(data, i, n)
	}

	// successively remove the max item and swap it to the end
	for i := n-1; i != -1; i-- {
		data[i], data[0] = data[0], data[i]
		trickleDown(data, 0, i)
	}
}

func testHeapsort(data []int) {
	timeFunction("ran heapsort", func() {
		heapsort(data);
	})
	if !sort.IntsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}


//--- mergesort: recursive implementation with two arrays

func mergesort(data, buffer []int, left, right int) {
	if left+1 == right {
		return
	}
	mid := (left+right)/2
	mergesort(data, buffer, left, mid)
	mergesort(data, buffer, mid, right)
	posLeft, posRight := left, mid
	posBuffer := left
	for {
		if data[posLeft] <= data[posRight] {
			buffer[posBuffer] = data[posLeft]
			posBuffer++
			posLeft++
			if posLeft == mid {
				for posRight != right {
					buffer[posBuffer] = data[posRight]
					posBuffer++
					posRight++
				}
				break
			}
		} else {
			buffer[posBuffer] = data[posRight]
			posBuffer++
			posRight++
			if posRight == right {
				for posLeft != mid {
					buffer[posBuffer] = data[posLeft]
					posBuffer++
					posLeft++
				}
				break
			}
		}
	}
	for i := left; i != right; i++ {
		data[i] = buffer[i]
	}
}

func testMergesort(data []int) {
	timeFunction("ran mergesort", func() {
		buffer := make([]int, len(data))
		mergesort(data, buffer, 0, len(data))
	})
	if !sort.IntsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}


//--- quicksort: in-place implementation with median-of-three pivot selection

func quicksortInt(data []int, left, right int) {
	if (left+1 >= right) {
		return
	}
	ai, bi, ci := left, (left+right)/2, right-1
	a, b, c := data[ai], data[bi], data[ci]
	var pos int
    if (a < b) {
        if (c < a) {
			pos = ai
		} else if (c < b) {
			pos = ci
		} else {
			pos = bi
		}
	} else {
        if (c < b) {
			pos = bi
		} else if (c < a) {
			pos = ci
		} else {
			pos = ai
		}
	}
	pivot := data[pos]
	data[pos] = data[right-1]
	tail := left
	for i := left; i != right-1; i++ {
		if (data[i] < pivot) {
			data[tail], data[i] = data[i], data[tail]
			tail++
		}
	}
	data[right-1] = data[tail]
	data[tail] = pivot
    quicksortInt(data, left, tail)
    quicksortInt(data, tail+1, right)
}

func testQuicksortInt(data []int) {
	timeFunction("ran quicksort", func() {
		quicksortInt(data, 0, len(data))
	})
	if !sort.IntsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}

func quicksortString(data []string, left, right int) {
	if (left+1 >= right) {
		return
	}
	ai, bi, ci := left, (left+right)/2, right-1
	a, b, c := data[ai], data[bi], data[ci]
	var pos int
    if (a < b) {
        if (c < a) {
			pos = ai
		} else if (c < b) {
			pos = ci
		} else {
			pos = bi
		}
	} else {
        if (c < b) {
			pos = bi
		} else if (c < a) {
			pos = ci
		} else {
			pos = ai
		}
	}
	pivot := data[pos]
	data[pos] = data[right-1]
	tail := left
	for i := left; i != right-1; i++ {
		if (data[i] < pivot) {
			data[tail], data[i] = data[i], data[tail]
			tail++
		}
	}
	data[right-1] = data[tail]
	data[tail] = pivot
    quicksortString(data, left, tail)
    quicksortString(data, tail+1, right)
}

func testQuicksortString(data []string) {
	timeFunction("ran quicksort", func() {
		quicksortString(data, 0, len(data))
	})
	if !sort.StringsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}

func quicksortStudent(data Students, left, right int) {
	if (left+1 >= right) {
		return
	}
	ai, bi, ci := left, (left+right)/2, right-1
	var pos int
    if (data.Less(ai, bi)) {
        if (data.Less(ci, ai)) {
			pos = ai
		} else if (data.Less(ci, bi)) {
			pos = ci
		} else {
			pos = bi
		}
	} else {
        if (data.Less(ci, bi)) {
			pos = bi
		} else if (data.Less(ci, ai)) {
			pos = ci
		} else {
			pos = ai
		}
	}
	pivot := data[pos]
	data[pos] = data[right-1]
	tail := left
	for i := left; i != right-1; i++ {
		if (data[i].LessThan(pivot)) {
			data[tail], data[i] = data[i], data[tail]
			tail++
		}
	}
	data[right-1] = data[tail]
	data[tail] = pivot
    quicksortStudent(data, left, tail)
    quicksortStudent(data, tail+1, right)
}

func testQuicksortStudent(data Students) {
	timeFunction("ran quicksort", func() {
		quicksortStudent(data, 0, len(data))
	})
	if !sort.IsSorted(sort.Interface(data)) {
		fmt.Printf("error: data is not sorted\n")
	}
}


//--- native sort

// for ints
func testNativeSortInteger(data []int) {
	timeFunction("ran native sort", func() {
		sort.Ints(data);
	})
	if !sort.IntsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}

// for strings
func testNativeSortString(data []string) {
	timeFunction("ran native sort", func() {
		sort.Strings(data);
	})
	if !sort.StringsAreSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}

// for anything that implements sort.Interface
func testNativeSort(data sort.Interface) {
	timeFunction("ran native sort", func() {
		sort.Sort(data);
	})
	if !sort.IsSorted(data) {
		fmt.Printf("error: data is not sorted\n")
	}
}


//--- student objects

type Student struct {
	Name string
	Grade int
}
type Students []Student

func (a Student) LessThan(b Student) bool {
	if a.Grade != b.Grade {
		return a.Grade < b.Grade
	}
	return a.Name < b.Name
}

// methods for sort.Interface
func (students Students) Len() int {
	return len(students)
}
func (students Students) Less(i, j int) bool {
	if students[i].Grade != students[j].Grade {
		return students[i].Grade < students[j].Grade
	}
	return students[i].Name < students[j].Name
}
func (students Students) Swap(i, j int) {
	students[i], students[j] = students[j], students[i]
}


//--- random permutation

func shuffle(data sort.Interface) {
    // constants for a linear congruential generator
	m := int64(1<<32)
	a := int64(1664525)
	c := int64(1013904223)
	randomValue := int64(42)  // seed
	for i := data.Len()-1; i != 0; i-- {
        randomValue = (a*randomValue + c) % m
		j := int(randomValue) % (i+1)
        if j != i {
            data.Swap(i, j)
        }
    }
}


//--- helper functions

func timeFunction(description string, fun func()) {
	startTime := time.Now()
	fun()
	elapsed := float64(time.Since(startTime))/1e9
	fmt.Printf("    %s  %.3f s\n", description, elapsed)
}

func displaySliceInt(data []int) {
	fmt.Printf("%d", data[0])
	for i := 1; i != len(data); i++ {
		fmt.Printf(" %d", data[i])
	}
	fmt.Printf("\n")
}

func displaySliceString(data []string) {
	fmt.Printf("%s", data[0])
	for i := 1; i != len(data); i++ {
		fmt.Printf(" %s", data[i])
	}
	fmt.Printf("\n")
}

func displaySliceStudent(data []Student) {
	fmt.Printf("%d %s", data[0].Grade, data[0].Name)
	for i := 1; i != len(data); i++ {
		fmt.Printf(" %d %s", data[i].Grade, data[i].Name)
	}
	fmt.Printf("\n")
}


func main() {
	if len(os.Args) != 2 {
		fmt.Printf("error: must specify the length of the array\n")
		return
	}
	n, _ := strconv.Atoi(os.Args[1])


	//--- integer sorting
	fmt.Println("integers:")

	// initialize an array with 0, ..., n-1
	var sortedIntegers []int
	timeFunction("made ordered integer array", func() {
		sortedIntegers = make([]int, n)
		for i := range sortedIntegers {
			sortedIntegers[i] = i
		}
	})
	//elapsed := float64(time.Since(startTime))/1e9
	//fmt.Printf("    made ordered integer array in %.3f s\n", elapsed)

	// permute randomly
	shuffledIntegers := make([]int, n)
	copy(shuffledIntegers, sortedIntegers)
	timeFunction("shuffled", func() {
		shuffle(sort.IntSlice(shuffledIntegers))
	})

	// sort integer arrays
	intData := make([]int, n)
	copy(intData, shuffledIntegers)
	testHeapsort(intData);
	copy(intData, shuffledIntegers)
	testMergesort(intData);
	copy(intData, shuffledIntegers)
	testQuicksortInt(intData);
	copy(intData, shuffledIntegers)
	testNativeSortInteger(intData);


	//--- string sorting
	fmt.Println("strings:")

    // make strings: 'a', 'b', ..., 'z', 'aa', 'ab', ..., 'az', 'ba', ...
	var sortedStrings []string
	timeFunction("made ordered string array", func() {
		sortedStrings = make([]string, n)
		chars := []byte{'a'}
		numChars := 1
		for i := 0; i != n; i++ {
			sortedStrings[i] = string(chars)
			pos := numChars-1
			for {
				if chars[pos] != 'z' {
					chars[pos]++
					break
				}
				pos -= 1
				if pos == -1 {
					chars = append([]byte{'a'}, chars...)
					numChars += 1
					break
				}
			}
			if pos != numChars-1 {
				for j := pos+1; j <= numChars-1; j++ {
					chars[j] = 'a'
				}
			}
		}
	})

    // permute randomly
	shuffledStrings := make([]string, n)
	copy(shuffledStrings, sortedStrings)
	timeFunction("shuffled", func() {
		shuffle(sort.StringSlice(shuffledStrings))
	})

	// sort string arrays
	stringData := make([]string, n)
	copy(stringData, shuffledStrings)
	testQuicksortString(stringData)
	copy(stringData, shuffledStrings)
	testNativeSortString(stringData)


    //--- object sorting
	fmt.Println("objects:")

    // make student objects based on the sorted strings
	var sortedStudents []Student
	timeFunction("made student objects", func() {
		sortedStudents = make([]Student, n)
		for i, s := range sortedStrings {
			sortedStudents[i] = Student{s, i%10}
		}
	})

    // permute randomly
	shuffledStudents := make([]Student, n)
	copy(shuffledStudents, sortedStudents)
	timeFunction("shuffled", func() {
		shuffle(Students(shuffledStudents))
	})

    // sort objects
	data := make([]Student, n)
	copy(data, shuffledStudents)
	testQuicksortStudent(Students(data))
	copy(data, shuffledStudents)
	testNativeSort(Students(data))

	/*
	timeFunction("ran native sort", func() {
		testQuicksort()
	})

    # extra test with sorting on keys
    data = shuffled_students[0..-1]
    time_block 'sorted natively with keys' do
        data.sort_by! {|student| [student.grade, student.name] }
    end
	*/
}
